### ISSUE TITLE
Split off some general purpose helper libraries

### ISSUE THREAD

**Issue by snoyberg** [commercialhaskell/stack#3620]
Discussed with @mgsloan last week. Now that the 1.6 release branch is cut, we can start to seriously consider it.

Basic idea: there are some subsets of functionality within the Stack codebase that are not Stack specific. There may be more, but at the least I see: general command line utility code (logging functions, pretty printer, `RIO` monad), and Cabal-related stuff (file parsing, index handling). We should split these off into separate packages.

Motivation: Doing this correctly will ensure that:

* We can reuse this functionality elsewhere (stackage-curator is a prime example)
* We're forced to define cleaner APIs with less breakage
* It becomes easier to test this subset of functionality
* We reduce the relative amount of code in Stack itself

Approach: doing this badly will greatly increase our maintenance burden on Stack. Instead of splitting off into separate packages that need to be release to Hackage and tracking with extra-deps, I recommend something along the lines of:

* Create a subdirectory inside the stack repo for holding these "split off" packages
* Start moving modules into packages within this new directory, but...
* Use `source-dirs` from the main `package.yaml` so that we don't need to disrupt the workflow of working on Stack
* When we're satisfied that these subpackages have a semi-stable API, we remove the `source-dirs` and start treating them as separate packages within the repo
* At the next release of Stack, we release all subpackages with the main package
* In the future, we may choose to split these packages off into separate repos if deemed stable enough

This is just a proposal for now, and certainly comes with some downside, so feedback from other contributors is strongly encouraged.

**Comment by lwm** [comment]
I'm in favour - I like the approach as well. Long term, once these packages go out, we'd reduce compile times, which would be really nice. Reduce LOC can help.

> We can reuse this functionality elsewhere (stackage-curator is a prime example)

This would be great.

> We're forced to define cleaner APIs with less breakage

Is this the beginning of a shift towards stable APIs in Stack? As far as I know, Stack has never offered a stable API.

> It becomes easier to test this subset of functionality

I'd question if we're not writing more tests because it is hard. I think no one has time ;) It'd certainly be nice to create more tests for specific moving parts and this will help.

**Comment by Blaisorblade** [comment]
👍 I suspect many of the .Extra packages are also candidates? Or are those a separate issue?

It would also be an excuse to rationalize some of those functions. Last I looked at this, I wished to rationalize the conduit wrappers used to launch subprocesses and capture IO.

I’m no conduit expert, but they seemed to be only subtly different, it was never clear to me which one to use and why they were all needed; I think I wondered at some point about potential bugs in the area. Probably there was none, but it was less than obvious.

**Comment by snoyberg** [comment]
Yes, definitely candidates, and yes, the process launching code should
definitely be cleaned up.

**Comment by mgsloan** [comment]
Once this stuff is in separate source-dirs, it'd also immediately be feasible to start using them in `stackage-curator`.  Just have stack as a submodule and add them to `stackage-curator`'s `hs-source-dirs`!  This could help with the process of refining the utilities.

**Comment by lwm** [comment]
So, since I think it's a general +1 all round - can someone lay out what to do? Perhaps split it off into different tickets, depending on how easy/hard they are, and see if people can pick them off in parallel?

**Comment by snoyberg** [comment]
I have this down as something to get started on myself at some point, most likely as I get deeper into the stackage-curator rewrite. I'm honestly not sure how I'm going to really attack this. If there are others that want to participate in the work with me, speak up here, and I'll try to figure out a better approach than "I'll do it when I can" :)

**Comment by snoyberg** [comment]
Next set of packages to split off:

* [ ] Something for parsing YAML with warnings, currently it's the `Data.Aeson.Extended` module, but that will likely change significantly
* [ ] rio-pretty: split off the pretty printing support present in Stack, and add direct support for generating `Utf8Builder` values instead of using `String`

Once those are done, I think it will be fair to close this issue and open up new issues for further splitting

**Comment by snoyberg** [comment]
These aren't going to happen immediately, closing as complete. We can open new issues in the future for further refactorings.
